File: karate-gatling/src/test/scala/mock/CatsChainedSimulation.scala
package mock

import com.intuit.karate.gatling.PreDef._
import io.gatling.core.Predef._

import scala.concurrent.duration._
import scala.util.Random

class CatsChainedSimulation extends Simulation {

  MockUtils.startServer(0)

  val protocol = karateProtocol(
    "/cats/{id}" -> Nil
  )

  val feeder = Iterator.continually(Map("name" -> (Random.alphanumeric.take(20).mkString + "-name")))

  val feederToKarate = scenario("feederToKarate")
    .exec(karateSet("name", session => session("name").as[String]))

  val create = scenario("create").exec(karateFeature("classpath:mock/cats-chained.feature@name=create"))

  val read = scenario("read").exec(karateFeature("classpath:mock/cats-chained.feature@name=read")).exec(session => {
    println("*** id in gatling: " + session("id").as[String])
    println("*** session status in gatling: " + session.status)
    session
  })

  val createAndRead = scenario("createAndRead").group("createAndRead") {
    feed(feeder)
      .exec(feederToKarate)
      .exec(create)
      // for demo: injecting a new variable name expected by the 'read' feature
      .exec(karateSet("expectedName", session => session("name").as[String]))
      .exec(read)
  }

  setUp(
    createAndRead.inject(rampUsers(10) during (5 seconds)).protocols(protocol)
  ).assertions(details("createAndRead").failedRequests.percent.is(0))

}

File: karate-gatling/src/test/java/mock/CatsChainedSimulation.java
package mock;

import java.util.Collections;
import java.util.Iterator;
import java.util.Map;
import java.util.Random;
import java.util.stream.Stream;
import static com.intuit.karate.gatling.javaapi.KarateDsl.*;
import static io.gatling.javaapi.core.CoreDsl.details;
import static io.gatling.javaapi.core.CoreDsl.feed;
import static io.gatling.javaapi.core.CoreDsl.rampUsers;
import static io.gatling.javaapi.core.CoreDsl.scenario;

import io.gatling.javaapi.core.*;

public class CatsChainedSimulation extends Simulation {

  public CatsChainedSimulation() {
    MockUtils.startServer(0);

    ProtocolBuilder protocol = karateProtocol(uri("/cats/{id}").nil());
  
    Iterator<Map<String, Object>> feeder = Stream.generate(() -> Collections.<String, Object>singletonMap("name", new Random().nextInt(20)+"-name")).iterator();
  
    ScenarioBuilder createAndRead = scenario("createAndRead")
    .group("createAndRead").on(
      feed(() -> feeder)
        .exec(karateSet("name", session -> session.getString("name")))
        .exec(karateFeature("classpath:mock/cats-chained.feature@name=create"))
        // for demo: injecting a new variable name expected by the 'read' feature
        .exec(karateSet("expectedName", session -> session.getString("name")))
        .exec(karateFeature("classpath:mock/cats-chained.feature@name=read")).exec(session -> {
      System.out.println("*** id in gatling: " + session.getString("id"));
      System.out.println("*** session status in gatling: " + session.asScala().status());
      return session;
    }));
    
  
    setUp(
      createAndRead.injectOpen(rampUsers(10).during(5)).protocols(protocol)
    ).assertions(details("createAndRead").failedRequests().percent().is(0d));
  
  }

}
